### ISSUE TITLE
Assessment instance list: Open (no time limit) inaccurate for PT linked exams

### ISSUE THREAD

**Issue by trombonekenny** [PrairieLearn/PrairieLearn#7267]
The instructor assessment instances page (Students tab) Remaining column is misleading about time limits when an exam is PrairieTest linked. See:
![image](https://user-images.githubusercontent.com/10765199/222218506-8dd4f0b0-aaad-4ded-947f-60b6645ea6de.png)

PT may in fact be enforcing a time limit, but PL's configuration doesn't know about that. It has confused a few instructors.

It would be nice if PL could figure out if the exam is PT enforced and put a note there (instead of `no time limit`), or even as far as showing the time limits from PT if we want to tightly integrate those.

**Comment by jonatanschroeder** [comment]
Maybe a note saying "Exam mode" would reduce the confusion in the meantime?

**Comment by trombonekenny** [comment]
> Maybe a note saying "Exam mode" would reduce the confusion in the meantime?

Except that "mode: Exam" is about access, not necessarily timing, and that's already confusing so I wouldn't want to overload things here. Unless we wanted to add something about "Expected auto-close at HH:MM" which could also be helpful. 

If we can identify that the assessment (or assessment instance) is having its time limits controlled by PT, something like `Open (PrairieTest enforced)` seems more accurate and points instructors to more information.

**Comment by jonatanschroeder** [comment]
There are ways to do this search now with a sproc call for each row, but that would probably cause performance issues. Identifying the mode used to open the assessment instance is trivial. So if the instance was started as mode exam and it is still open, I'd assume that it's almost always "PrairieTest enforced", though there are corner cases where this would not be true. What I can say is that PrairieTest was likely involved in this assessment at some point, which is why I suggested "exam mode" or another expression in the same vein.

**Comment by trombonekenny** [comment]
You would know that PT is involved in the assessment in the allowAccess, which might be "good enough" to not query per instance (but I don't think adding some PT joins to the per row sprocs would be too bad since we're already doing a query for each anyway?)

**Comment by jonatanschroeder** [comment]
I'm just worried of cases (including some used by @firasm ) that have a combination of Exam mode and regular mode in some capacity. The fact that there is an exam-mode rule does not necessarily mean that this particular student is using PT. And even if there is a PT record for the student, what happens with students that leave the CBTF without formally closing the assessment? The assessment will remain open, though (presumably) not accessible outside of that environment. Auto-close will probably kick-in to close the assessment, but that takes time.

I know these are corner cases, and not common, but I'm a bit wary of providing information that may not be accurate.

**Comment by trombonekenny** [comment]
Then it sounds like you're advocating for per-row inspection for the timing?

**Comment by jonatanschroeder** [comment]
> Then it sounds like you're advocating for per-row inspection for the timing?

Not necessarily. I'm just advocating for caution in presenting information to make sure it is accurate. If precise information cannot be presented accurately without a significant loss of performance, maybe the information should be vague enough to be correct but still provide meaningful information. This is why I suggested the "exam mode" note, since that is accurate (the student indeed started the assessment in exam mode) and does not incur a performance penalty (since we are using information in a table that is already queried). Accurately querying the PT exam is a bit more complicated, since the link between PL assessment instances and PT reservations is based on exam UUIDs in access rules, which may not trivial to query for a large number of students.

**Comment by firasm** [comment]
I sometimes get confused about the information on the "Students" page, and then I remember that it's because I'm doing silly things like having access rules that allow some students to do the test on PL, and the majority to do it on PT.

Without knowing the full details of the implications, one idea I had to resolve what I think the information display part of the problem is to lay the information out in "blocks"

Section 1: PrairieLearn Exam
- Students who are writing using the PL environment 

Section 2: PrairieTest Exam
- Students who did the test in a PrairieTest environment

That way, we could at least separate out the instances from the environment they did it in so it's clearer how they did it (which is usually what I look for). Then you could also provide information that's guaranteed to be correct, including "Remaining time is not available because this is a PT instance" if it's too expensive to query that information.

**Comment by nwalters512** [comment]
AFAIK we don't know/track if something is a "PT instance". I think the best we can do is check at the current instance in time whether an exam's access is being controlled by PT, but that's potentially expensive.

**Comment by jonatanschroeder** [comment]
> AFAIK we don't know/track if something is a "PT instance". I think the best we can do is check at the current instance in time whether an exam's access is being controlled by PT, but that's potentially expensive.

I believe the `assessment_instances.mode` field indicates the mode used when the instance was created.
